File: apps/miniapp/src/services/ai.ts
import { request } from "./http";

export type AIReferenceKind = "PLOT" | "PRODUCTION" | "OPERATION" | "HARVEST" | "SPECIES";

export interface AIReference {
  kind: AIReferenceKind;
  id: number;
  label: string;
  route: string;
}

export interface AITurnInput {
  conversationId: number;
  message: string;
}

export interface AITurnResult {
  answer: string;
  references: AIReference[];
  needsSelection: boolean;
  candidates: AIReference[];
}

export interface AIConversation {
  id: number;
  farmId: number;
  farmName: string;
  title: string;
  createdAt: string;
  updatedAt: string;
}

export interface AIConversationPage {
  items: AIConversation[];
  page: number;
  pageSize: number;
  total: number;
}

export interface AIMessage {
  id: number;
  role: "user" | "assistant";
  content: string;
  references: AIReference[];
  candidates: AIReference[];
  createdAt: string;
}

export interface AIMessagePage {
  items: AIMessage[];
  page: number;
  pageSize: number;
  total: number;
}

export function createAIConversation(farmId: number): Promise<AIConversation> {
  return request<AIConversation>({
    url: "/ai/conversations",
    method: "POST",
    data: { farmId },
  });
}

export function getAIConversations(page = 1, pageSize = 20): Promise<AIConversationPage> {
  return request<AIConversationPage>({
    url: `/ai/conversations?page=${page}&pageSize=${pageSize}`,
  });
}

export function getAIMessages(
  conversationId: number,
  page = 1,
  pageSize = 100,
): Promise<AIMessagePage> {
  return request<AIMessagePage>({
    url: `/ai/conversations/${conversationId}/messages?page=${page}&pageSize=${pageSize}`,
  });
}

export function deleteAIConversation(conversationId: number): Promise<void> {
  return request<void>({
    url: `/ai/conversations/${conversationId}`,
    method: "DELETE",
  });
}

export function createAITurn(input: AITurnInput): Promise<AITurnResult> {
  return request<AITurnResult>({
    url: "/ai/turn",
    method: "POST",
    data: input,
  });
}

File: apps/api/app/schemas/ai.py
from datetime import datetime
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.config import settings

MAX_HISTORY_MESSAGES = settings.ai_max_history_turns * 2


class CreateAIConversationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    farm_id: int = Field(
        gt=0,
        validation_alias="farmId",
        serialization_alias="farmId",
    )


class AITurnRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    conversation_id: int = Field(
        gt=0,
        validation_alias="conversationId",
        serialization_alias="conversationId",
    )
    message: Annotated[str, Field(min_length=1, max_length=settings.ai_max_message_chars)]

    @field_validator("message")
    @classmethod
    def validate_message(cls, value: str) -> str:
        message = value.strip()
        if not message:
            raise ValueError("message cannot be blank.")
        return message

class AIReference(BaseModel):
    kind: Literal["PLOT", "PRODUCTION", "OPERATION", "HARVEST", "SPECIES"]
    id: int
    label: str
    route: str


class AITurnResponse(BaseModel):
    answer: str
    references: list[AIReference] = Field(default_factory=list)
    needs_selection: bool = Field(default=False, serialization_alias="needsSelection")
    candidates: list[AIReference] = Field(default_factory=list)


class AIConversationResponse(BaseModel):
    id: int
    farm_id: int = Field(serialization_alias="farmId")
    farm_name: str = Field(serialization_alias="farmName")
    title: str
    created_at: datetime = Field(serialization_alias="createdAt")
    updated_at: datetime = Field(serialization_alias="updatedAt")


class AIConversationPage(BaseModel):
    items: list[AIConversationResponse]
    page: int
    page_size: int = Field(serialization_alias="pageSize")
    total: int


class AIMessageResponse(BaseModel):
    id: int
    role: Literal["user", "assistant"]
    content: str
    references: list[AIReference] = Field(default_factory=list)
    candidates: list[AIReference] = Field(default_factory=list)
    created_at: datetime = Field(serialization_alias="createdAt")


class AIMessagePage(BaseModel):
    items: list[AIMessageResponse]
    page: int
    page_size: int = Field(serialization_alias="pageSize")
    total: int
